Separate mountingSystem definition from instance

grimsa opened the pull request:
This PR separates `mountingSystem` into:
- `mountingSystemDefinition` (holding attributes common to this type of mounting system) and
- `mountingSystem` (holding attributes related to a specific instance).

This pattern is all over IEP model. The goal of this change is to make further changes/improvements to mounting system model easier.

**Element structure after proposed change:**

    - pvSystem
        - mountingSystemDefinitions
        - pvArrays
            - pvArray
                - mountingSystem
            - pvArray
                - mountingSystem

**Attributes on definition and instance**
Overall I did not add/remove any attributes, they are just moved to either definition or instance.
Let me know if I should move any of them to the other type.

   - mountingSystemDefinition
       - id (attribute - ID)
       - manufacturer (string)
       - componentsDescription (string)
       - numberOfRowsPerRack (int)
       - anchorMaxHorizontalSpacing (double)
       - anchorFastenersMeetManufacturerGuidelines (boolean)
       - anchorFastenerDiameter (double)
       - anchorFastenerEmbedmentDepth (double)
       - anchorFastenersPerAnchor (int)
       - roofPenetrationWeatherProofing (string)
   - mountingSystem
       - mountingSystemDefinitionIdRef (attribute - IDREF)
       - anchorsInAdjacentRowsStaggered (boolean)
       - totalArrayWeight (double)
       - totalPanelSurfaceArea (double)

grimsa committed:
Separate mountingSystem definition from instance

grimsa commented:
Alternative element structure is also a possibility - let me know if you'd prefer this one:

    - pvSystem
        - mountingSystemDefinitions
        - mountingSystems
            - mountingSystem
            - mountingSystem
        - pvArrays
            - pvArray
                - mountingSystemIdRef
            - pvArray
                - mountingSystemIdRef

mpalmquist commented:
I have no issues with implementing this pattern for mounting systems. However, I do not know if the attributes are properly distributed as to what is part of the definition and what attributes are specific to an instance. Has Derek reviewed these?

ceniza commented:
Looking at your proposal, there are a few attributes listed under MountingSystemDefinition that belong in Mounting System.

   - mountingSystemDefinition
       - id (attribute - ID)
       - manufacturer (string)
       - componentsDescription (string)
       - anchorFastenerDiameter (double) 
       - roofPenetrationWeatherProofing (string)
   - mountingSystem
       - mountingSystemDefinitionIdRef (attribute - IDREF)
       - anchorsInAdjacentRowsStaggered (boolean)
       - totalArrayWeight (double)
       - totalPanelSurfaceArea (double)
       - numberOfRowsPerRack (int) - will vary by instance
       - anchorMaxHorizontalSpacing (double) - will vary by instance
       - anchorFastenersMeetManufacturerGuidelines (boolean) - vary by instance. 
       - anchorFastenerEmbedmentDepth (double) - will vary by instance in some cases
       - anchorFastenersPerAnchor (int) - will vary by instance in some cases

I think we should add the following to mountingSystemDefinition
   - minAnchorFastenersPerAnchor (int) - usually 1, but sometimes 1 or 2
   - maxAnchorFastenersPerAnchor (int) - usually 1, but sometimes up to 8
   - railManufacturer (string) - rail and anchor manufacturers can be different
   - railProductFamily (string)
   - anchorManufacturer (string) - rail and anchor manufacturers can be different
   - anchorModel (string) 

Where is commonModuleStandoff currently? (It should be in the instance. A more detailed data model would derive this with calculations based on the dimensions of the anchor and rail, but we don't need that now.)

These above will be needed. The currently functional request looks like
`
    <iep:mountingSystem id="null-135663">
      <iep:name>82213366-7011-4e88-a056-d89f177d42bf</iep:name>
      <iep:railManufacturer>Unirac</iep:railManufacturer>
      <iep:railProductFamily>SolarMount</iep:railProductFamily>
      <iep:anchorManufacturer>Unirac</iep:anchorManufacturer>
      <iep:anchorModel>Flashkit Pro</iep:anchorModel>
      <iep:numberOfRowsPerRack>1</iep:numberOfRowsPerRack>
      <iep:anchorMaxHorizontalSpacing>1.22</iep:anchorMaxHorizontalSpacing>
      <iep:totalArrayWeight>319.2</iep:totalArrayWeight>
      <iep:totalPanelSurfaceArea>31.2</iep:totalPanelSurfaceArea>
      <iep:anchorsInAdjacentRowsStaggered>true</iep:anchorsInAdjacentRowsStaggered>
      <iep:commonModuleStandoff>0.0762</iep:commonModuleStandoff>
    </iep:mountingSystem>
`

I think this would be the minimum for an integration at this point.

mpalmquist commented:
Once this PR is updated with Derek's suggested changes I'll accept it.

On Wed, Nov 14, 2018 at 9:26 PM ceniza <notifications@github.com> wrote:

> Looking at your proposal, there are a few attributes listed under
> MountingSystemDefinition that belong in Mounting System.
>
>    - mountingSystemDefinition
>       - id (attribute - ID)
>       - manufacturer (string)
>       - componentsDescription (string)
>       - anchorFastenerDiameter (double)
>       - roofPenetrationWeatherProofing (string)
>    - mountingSystem
>       - mountingSystemDefinitionIdRef (attribute - IDREF)
>       - anchorsInAdjacentRowsStaggered (boolean)
>       - totalArrayWeight (double)
>       - totalPanelSurfaceArea (double)
>       - numberOfRowsPerRack (int) - will vary by instance
>       - anchorMaxHorizontalSpacing (double) - will vary by instance
>       - anchorFastenersMeetManufacturerGuidelines (boolean) - vary by
>       instance.
>       - anchorFastenerEmbedmentDepth (double) - will vary by instance in
>       some cases
>       - anchorFastenersPerAnchor (int) - will vary by instance in some
>       cases
>
> I think we should add the following to mountingSystemDefinition
>
>    - minAnchorFastenersPerAnchor (int) - usually 1, but sometimes 1 or 2
>    - maxAnchorFastenersPerAnchor (int) - usually 1, but sometimes up to 8
>    - railManufacturer (string) - rail and anchor manufacturers can be
>    different
>    - railProductFamily (string)
>    - anchorManufacturer (string) - rail and anchor manufacturers can be
>    different
>    - anchorModel (string)
>
> Where is commonModuleStandoff currently? (It should be in the instance. A
> more detailed data model would derive this with calculations based on the
> dimensions of the anchor and rail, but we don't need that now.)
>
> These above will be needed. The currently functional request looks like
> <iep:mountingSystem id="null-135663">
> <iep:name>82213366-7011-4e88-a056-d89f177d42bf</iep:name>
> <iep:railManufacturer>Unirac</iep:railManufacturer>
> <iep:railProductFamily>SolarMount</iep:railProductFamily>
> <iep:anchorManufacturer>Unirac</iep:anchorManufacturer>
> <iep:anchorModel>Flashkit Pro</iep:anchorModel>
> <iep:numberOfRowsPerRack>1</iep:numberOfRowsPerRack>
> <iep:anchorMaxHorizontalSpacing>1.22</iep:anchorMaxHorizontalSpacing>
> <iep:totalArrayWeight>319.2</iep:totalArrayWeight>
> <iep:totalPanelSurfaceArea>31.2</iep:totalPanelSurfaceArea>
> <iep:anchorsInAdjacentRowsStaggered>true</iep:anchorsInAdjacentRowsStaggered>
> <iep:commonModuleStandoff>0.0762</iep:commonModuleStandoff>
> </iep:mountingSystem>
>
> I think this would be the minimum for an integration at this point.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/mpalmquist/IEPModel/pull/10#issuecomment-438922149>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AATwP-Y7S7idRMTh_f8S-ZE67nrGxBh6ks5uvPrvgaJpZM4X9Xal>
> .
>


-- 
Michael Palmquist

*SolarNexus <https://solarnexus.com>510.508.7697*

grimsa committed:
Add new elements to mountingSystemDefinition

grimsa commented:
Updated first commit according to comments and added another one which adds new elements to `mountingSystemDefinition`.

**Attributes on definition and instance in latest version**

   - mountingSystemDefinition
       - id (attribute - ID)
       - manufacturer (string)
       - componentsDescription (string)
       - railManufacturer (string, **new**)
       - railProductFamily (string, **new**)
       - anchorManufacturer (string, **new**)
       - anchorModel (string, **new**)
       - minAnchorFastenersPerAnchor (integer, **new**)
       - maxAnchorFastenersPerAnchor (integer, **new**)
       - anchorFastenerDiameter (double)
       - roofPenetrationWeatherProofing (string)

   - mountingSystem
       - mountingSystemDefinitionIdRef (attribute - IDREF)
       - anchorTotal (int, **old element - missed this in the previous list of attributes**)
       - anchorsInAdjacentRowsStaggered (boolean)
       - anchorMaxHorizontalSpacing (double)
       - anchorFastenersMeetManufacturerGuidelines (boolean)
       - anchorFastenerEmbedmentDepth (double)
       - anchorFastenersPerAnchor (int)
       - numberOfRowsPerRack (int)
       - totalArrayWeight (double)
       - totalPanelSurfaceArea (double)

--
Regarding `commonModuleStandoff` - currently IEP model has it as `pvArray/commonModuleStandoff` (double).

ceniza commented:
@grimsa @mpalmquist  Common module standoff is going to be a calculated value based on the height of the mount and rail. We are not including these details in the model right now, so perhaps it should be a MountingSystem property.
